File: HelloAumazondible/app/src/main/java/com/dstrube/helloaumazondible/MyAdapter.java
package com.dstrube.helloaumazondible;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.BaseAdapter;

import com.android.volley.toolbox.NetworkImageView;

import java.util.ArrayList;

public class MyAdapter extends BaseAdapter implements MyDataSource.OnDataChangedListener,
        AbsListView.OnScrollListener {

    private final ArrayList<String> imagesUrls;
    private final int screenSize;
    private final Context context;
    private final MyDataSource dataSource;

    public MyAdapter(Context context, MyDataSource dataSource, int screenSize) {
        super();
        this.screenSize = screenSize;
        this.dataSource = dataSource;
        this.context = context;
        imagesUrls = dataSource.getImagesUrls();
        this.dataSource.setOnDataChangedListener(this);
    }

    public void setImageSize(NetworkImageView networkImageView, int pos) {
        int size = screenSize;
        if (imageIsSmall(pos))
            size = size / 2;
        size = size - (int) (size * .05);
        networkImageView.setMaxHeight(size);
        networkImageView.setMinimumHeight(size);
        networkImageView.setMinimumWidth(size);
        networkImageView.setMinimumHeight(size);
    }

    public boolean imageIsSmall(int pos) {
        return pos % 3 != 0;
    }

    //
    // Overriding adapter methods
    //
    @Override
    public int getCount() {
        if (imagesUrls != null) {
            return imagesUrls.size();
        }
        return 0;
    }

    @Override
    public Object getItem(int pos) {
        if (imagesUrls != null) {
            return imagesUrls.get(pos);
        }
        return null;
    }

    @Override
    public long getItemId(int pos) {
        return 0;
    }

    @Override
    public View getView(int pos, View convertView, ViewGroup parent) {
        // volley NetworkImageView
        NetworkImageView networkImageView;
        if (convertView != null) {
            networkImageView = (NetworkImageView) convertView;
        } else {
            networkImageView = new NetworkImageView(this.context);
        }
        setImageSize(networkImageView, pos);
        networkImageView.setImageUrl(imagesUrls.get(pos),
                MyApplication.getImageLoader());
        return networkImageView;
    }

    @Override
    public void onDataChanged() {
        notifyDataSetChanged();
    }

    @Override
    public void onScrollStateChanged(AbsListView view, int scrollState) {
    }

    @Override
    public void onScroll(AbsListView view, int firstVisibleItem,
                         int visibleItemCount, int totalItemCount) {
        int count = getCount();
        if (count != 0 && firstVisibleItem > count - 10 && !dataSource.isLoading()) {
            dataSource.getMoreData();
        }
    }
}

File: kotlin/fix/HelloAumazondible/app/src/main/java/com/dstrube/helloaumazondible/MyAdapter.kt
package com.dstrube.helloaumazondible

import android.content.Context
import android.view.View
import android.widget.AbsListView
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.android.volley.toolbox.NetworkImageView


class MyAdapter(private val context: Context, private val dataSource: MyDataSource, private val screenSize: Int)
    : BaseAdapter(), MyDataSource.OnDataChangedListener, AbsListView.OnScrollListener {

    private val imagesUrls: ArrayList<String>?

    init {
        imagesUrls = dataSource.imagesUrls
        this.dataSource.setOnDataChangedListener(this)
    }

    fun setImageSize(networkImageView: NetworkImageView, pos: Int) {
        var size = screenSize
        if (imageIsSmall(pos))
            size = size / 2
        size = size - (size * .05).toInt()
        networkImageView.setMaxHeight(size)
        networkImageView.setMinimumHeight(size)
        networkImageView.setMinimumWidth(size)
        networkImageView.setMinimumHeight(size)
    }

    fun imageIsSmall(pos: Int): Boolean {
        //elegant, but harder to debug:
//        return if (pos % 3 == 0) false else true
        if (pos % 3 == 0){
            return false
        }
        else {
            return true
        }
    }

    //
    // Overriding adapter methods
    //
    override fun getCount(): Int {
        return imagesUrls?.size ?: 0
    }

    override fun getItem(pos: Int): Any? {
        if (imagesUrls != null) {
            return imagesUrls[pos]
        } else {
            return null
        }
    }

    override fun getItemId(pos: Int): Long {
        return 0
    }

    override fun getView(pos: Int, convertView: View?, parent: ViewGroup): View {
        // volley NetworkImageView
        val networkImageView: NetworkImageView
        if (convertView != null) {
            networkImageView = convertView as NetworkImageView
        } else {
            networkImageView = NetworkImageView(this.context)
        }
        setImageSize(networkImageView, pos)
        networkImageView.setImageUrl(imagesUrls!![pos],
                MyApplication.getImageLoader())
        return networkImageView
    }

    override fun onDataChanged() {
        notifyDataSetChanged()
    }

    override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {}

    override fun onScroll(view: AbsListView, firstVisibleItem: Int,
                 visibleItemCount: Int, totalItemCount: Int) {
        val count = count
        if (count == 0) {
            return
        } else if (firstVisibleItem > count - 10 && !dataSource.isLoading) {
            dataSource.getMoreData()
        }
    }
}
